Warping crossbreed 2 : now with 25% less communism

Caldony opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
Warping crossbreeds are here...again. Each warping crossbreed works like a paintbrush able to draw a "rune" unique to them. Each rune has a shitty MS paint coder sprite from yours truly, sorry about that I can't draw.

You can then "store" the rune back into your crossbreed to redraw it somewhere else, making them very easy to transport, some runes have more "charges" than others meaning you can draw more than one rune at a time.

 Some runes take longer to draw and store than others depending on their power. Rune can also be really easily destroyed by cleaning them, they can also be destroyed much faster with a bluespace crystal if you somehow can't afford a mop.

Anyway here's a long ass list on what each rune does:

- Grey rune: for 8 slime extract of the same type on the rune it will produce an adult slime of the same color of said extract. Useful to speed up crossbreed production. 

- Orange rune: Will put anyone who walk on the rune on fire, as long as the person that stepped on the rune is on fire they will keep teleporting every 7 seconds.

- Purple rune: Will transform plastic on the rune into regenerative mesh and cloth into suture. One of the most helpful runes for medical

- Blue rune: Constantly creates lube over itself as long as the rune is active. Good if you hate seeing people happy and fulfilled

- Metal rune: Act as an impenetrable wall. The only way to bypass it is by destroying it. Basically like the cultist barrier rune but constant and easier to destroy.

- Yellow rune: Will constantly recharge the battery of the room its in including the battery of any battery fueled things on it like guns/batons. good if you want to keep a specific room running at full power at all times.

- Dark purple rune: Will fill any beakers on the rune with plasma dust. Only really useful for xenobiologist.

- Dark blue rune: Will make anyone that walk on the rune very cold (-1000° to be exact) Usually won't kill anyone unless they are already weak to cold or they stay on the rune for too long.

- Silver rune: Will absorb any nutrients from food on the rune, then will redistribute the nutrients to the next person that decide to stay on the rune. 

- Bluespace rune: Will swap the place of the next two people that walk on the rune like a quantum spin inverter. Has a slight cooldown to avoid lag abuse.

- Sepia rune: Anyone walking on the rune will become slower for a little while and older, too old and you become nearsighted, WAY too old and you get dusted. 

- Cerulean rune: Will make a hologram of the latest mob that passed by it, the hologram will say the last 10 sentences the mob has said at random for as long as it lives.

- Pyrite rune: Randomly colors whatever steps on the rune, that's it, the color can still be washed off.

- Red rune: Anyone on the rune will deal cellular damage instead of brute damage per punch. if you punch something sentient each punch will also heal you for brute and burn damage. Also comes in with a special edgy hate fire effect behind you.

- Green rune: Transforms plasma into xeno resin, which you can use to make basic xenomorph structure. Mainly to make the crew freak out over fake xenomorph.

- Pink rune: gives you a hug that makes you happy UwU . Good if you want to run the psychiatrist out of business with artificial happiness

- (reworked)Gold rune: absorb hostile simple mobs over the rune. The next person that walk on the rune will release the mobs from the rune and be the priority target of said mobs. Can contain up to 5 mobs. One of the most lethal runes to step on depending on the mobs stored by the rune.

- Adamantine rune: Will copy ores in a 7 block radius from nearby mineral rocks. Obviously only works on mining places. Here if miners can't be bothered doing their fucking job or to mass-produce rare ores.

- Lightpink rune:When used, it will resurrect a ghostless/mindless person at full health by yeeting a (willing) random ghost into said body. Good to save "lost cause" corpses.

- Black rune: Will swap the species of the last two people that walked on the rune.

- Oil rune: Will protect anyone on the rune from outside explosions no matter the actual power of the explosion, Good if you want to see the damage of your maxcap first hand or survive a nuke.

- Rainbow rune: Will create your own little hotel room that you can hang out in! the room won't be destroyed until both the room is empty of any "customers" and the rune is destroyed. anything left in the room before that may or may not be deleted after the room is destroyed.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
Crossbreeds haven't been updated in a long time, and warping crossbreed adds unique tools to the xenobiologist arsenal that can both help and hinder the station depending on how they use it. 

Some of the runes can seem overpowered but it's good to remember that they are static and usually really easy to see, meaning you need quite a lot of preparation and luck to truly surprise someone with it.
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl: Caluan
add: Nanotrasen has finally found the secrets to the elusive bluespace crossbreed! Time to experiment!
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Caldony committed:
Adds the warping crossbreed!

all the warping crossbreeds are here! and they work! I still have very shitty MS paint coder sprites for each of them so sorry about that.

fludd12 commented:
im sorry for my continued absence, thank you for putting work into this ♥

enkidienne commented:
i wonderfully need that PR to pass

XDTM commented:
Full Opinions:

**Grey:** mundane utility, so fits nicely for grey
**Orange:** The teleport is kinda random, i think that firestacks + ignite is effective enough for a tier 1 color. Maybe adding some direct fire damage instead so it's pseudo-lava.
**Purple:** I feel this kinda wastes the 'regeneration' theme of purple; it could easily be some kind of healing rune instead.
**Blue:** very good choice
**Metal:** the barrier part kinda fits the theme, but i feel like it should not be invisible. Maybe summoning an instant reinforced wall, or something like that?
**Yellow**: the way it's coded makes it quite expensive in processing time (and also quite powerful). I'd suggest, as an alternative, making it recharge all items only on top of it (including equipped items), but faster.
**Dark Purple**: a bit too niche in my opinion. Personally I'd make it a 'trap' rune that injects random toxins onto people that step on it. It could also be more elaborate, where it could be splashed with toxins to select which toxins they inject.
**Dark Blue**: good as it is, but it could go a step further and inflict the freon freezing debuff (on a cooldown).
**Silver**: Could be simpler and simply fill nutrition, since anyone making that might as well eat/serve the food they plan on feeding to the rune.
**Bluespace**: Probably a bit too niche and unreliable to be useful. I'd make the extract draw two paired teleport runes that act as a permanent portal, kinda like a quantum pad but faster to use and easier to set up.
**Sepia**: i'd avoid using age as a game mechanic. It could simply slow, or work as a timestop trap, or be activatable for a wider timestop, or something else like that instead.
**Cerulean**: can't really see a proper use for it, to be honest. Then again, cerulean kinda suffers from not having a properly defined theme around it in general.
**Pyrite**: Fits the theme nicely
**Red**: Seems cool in concept, but kinda doesn't work when your goal when fighting is to down the other party as soon as possible, not lifestealing or causing long-lasting damage. 
**Green**: The effect fits the green theme, but could also easily be a handheld extract from some other crossbreed. I'd propose some kind of mutation-inducing rune in its place.
**Pink**: fits nicely enough with the mood theme.
**Gold**: material transmutations have caused trouble with the economy in the past, i'd suggest to avoid this kind of mechanic for the time being. Plus, gold slimes are more geared towards a summoning theme, which could work nicely as a rune.
**Adamantine**: Probably a bit overpowered, but it's the kind of overpowered that not many people will bother doing, so it's fine.
**Lightpink**: I think it's fine, but i'd remove the healing and keep the ghost infusion, so it's pretty much a method to insert more people into the round with no strings attached.
**Black**: Very unbalanced if you get your hands on a plasmaman corpse.
**Oil**: More unique than the landmine effect that i would've given it. Possibly a bit too niche, but i could see people nuking the shuttle and ending up standing on a free-floating floor tile until they reach centcomm.
**Rainbow**: Very unique, and sounds fun to play around with. Kinda erodes the hillbert's hotel's uniqueness, but rainbow slimes are pretty hard to get too.

Caldony committed:
Updates a few crossbreed

-Silver will now directly fill your nutrition for one "stack" of nutrition per 5 nutriment in the rune. 
-Orange no longer teleports you but gives you more fire stacks than before.
-Black rune now needs the person to be alive in order for the species swap to happen
(gold will probably be changed in the near future)

evilcyborgmrt commented:
Damn xenobio always gets the cool shit

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+	var/mob/living/carbon/human/slowed_down
+	if(istype(crossing,/mob/living/carbon/human))
+		slowed_down = crossing
+		slowed_down.add_movespeed_modifier(/datum/movespeed_modifier/status_effect/sepia_rune)
```
if you walk across this rune, then walk back across it while you're still slow, the slowdown effect won't be renewed because of this.

but if you walk across the rune, walk back across it before the slowdown effect expires, wait for the slowdown effect to expire, then walk back across the rune, the new slowdown effect will end "early" when the slowdown effect from the second crossing "should" have ended.

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+/*Cerulean crossbreed : creates a hologram of the last person that stepped on the tile */
+
+
+/obj/item/slimecross/warping/cerulean
```
another really cool rune, good job on this one

not only can you use it to see who last tried to break into your department, but you can also use this to interrogate corpses (even if they're braindead)

good shit right here, definitely gonna be fun to mess around with as a detective or something

XDTM replied:
i'm having a hard time trying to figure out any potential use case for this

the kind of person that talks about their crimes instead of doing them silently is also talkative enough to accumulate more than 10 messages before being caught

ATH1909 replied:
nah, you put the VICTIM on the rune

so that you can learn their last words, even if they went SSD or took a ghost role or something

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+		return
+	cooldown = max_cooldown + world.time
+	if(holotile && length(recent_speech))
+		holotile.say(recent_speech[pick(recent_speech)]) //say one of the 5 latest sentence said by the holo_host
```
update this comment

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+	holotile.name = "[holo_host.name] (Hologram)"
+	holotile.add_atom_colour("#77abff", FIXED_COLOUR_PRIORITY)
+	holotile.copy_overlays(holo_host, TRUE)
+
```
make this reset the cooldown as well, both to stop lag-spamming by putting a shitton of monkeys on a conveyorbelt or somethign and to make it less apparent to the person who just walked across the rune that the rune is gonna be repeating their speech once every 10 seconds

optimumtact commented:
My opinon:
The slime crossbreed proliferation is shit

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+	return ..()
+
+
+/obj/item/slimecross/warping/pyrite
```
crayons are back, baby

decent rune, can be useful for both griefing and powergaming (time to grind for a dark sprite!)

XDTM replied:
good point, probably better to make it a list of colors instead of fully random to prevent that kind of powergaming

ATH1909 replied:
eh, I personally see it as a feature

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+///colors whatever steps on the rune randomly
+/obj/effect/warped_rune/pyritespace/Crossed(atom/movable/AM)
+	. = ..()
+	AM.color = rgb(rand(0,255),rand(0,255),rand(0,255))
```
is it okay that this will recolor ANYTHING that crosses it, not just living creatures?

it's probably fine, but I thought that you should be made aware of this.

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+/* Will make anyone on the rune do genetic punch damage. May need rebalancing. Not compatible with gloves of the north star*/
+
+
+/obj/item/slimecross/warping/red
```
awful, AWFUL rune

literally bottom tier

not useful for griefing people, killing people, OR helping people

literally when would you ever use this

ATH1909 replied:
if you REALLY want to stick with the "soul-punching fists" idea, AT LEAST make the rune make you immovable while you're alive and on it

ATH1909 replied:
why not try something like the blood beam ability of the blood rites cult spell? that ability's cool, but it's so impractical that we barely ever see it (but hey, it's more practical than this rune in its current state)

ATH1909 replied:
or maybe just cut the soul and fist stuff and just add an effect that gives all of your attacks (including your ranged attacks) lifesteal

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+	effect_desc = "Makes a rune that will periodically create plasma dust,to harvest it simply put a beaker of some kind over the rune."
+
+
+/obj/effect/warped_rune/darkpurplespace
```
it'd be nice if you could let this interact with plumbing somehow, so that chemists could use this for some plumbing setups or something

ATH1909 replied:
being able to just scoop or syringe plasma out of this rune (maybe give it a reagent capacity that slowly fills itself up with plasma?) would also be nice

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+/obj/effect/warped_rune/greenspace/proc/transmute_resin()
+	for(var/obj/item/stack/sheet/mineral/plasma/plasma_sheet in rune_turf)
+		resin = new(rune_turf)
+		plasma_sheet.amount--
```
same comment here that I had with the plastic and cloth conversion rune

XDTM replied:
is there any use for resin for non-aliens?

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+/* Green rune vores plasma and spews out xeno resin which lets you build xeno structure*/
+
+
+/obj/item/slimecross/warping/green
```
pretty cool idea for a rune

it'd be nice if you could make xeno organs (aside from the neurotoxin spitter and the eggsac (the other human-like xeno organs (xeno eyes, xeno tongue, etc.) and the xeno-unique organs (hive node, plasma vessel, resin spitter, etc.) should be fine, though)) with the resin from it

imagine medbay using xeno organs as replacement organs instead of cybernetic ones, that'd be badass

XDTM replied:
something that unique would deserve a proper progression path to get, not a random crossbreed reaction

ATH1909 replied:
you keep live xenos around for organ harvesting for the really good organs, like the neurotoxin one and the egg sac

still, I could see an argument for only allowing for the crafting of xeno organs that have human-equivalents (xeno eyes, xeno livers, etc.)

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+//note : some of these can only be built ON resin weeds such as the resin nest.
+GLOBAL_LIST_INIT(resin_recipes, list ( \
+	new/datum/stack_recipe("Resin seed", /obj/structure/alien/weeds/node, 10, one_per_turf = 1, on_floor = 1, time = 100), \
+	new/datum/stack_recipe("Resin Wall", /obj/structure/alien/resin/wall, 3, one_per_turf = 1, on_floor = 1 , time = 50), \
```
don't forget about resin windows

ATH1909 replied:
these are all kinda expensive, seeing as how you need an entire plasma sheet for each resin sheet

either lower the costs of these recipes, or increase amount of resin sheets that you get per plasma sheet

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+/*People who step on the dark blue rune will suddendly get very cold,pretty straight forward.*/
+
+
+/obj/item/slimecross/warping/darkblue
```
being able to be hooked into plumbing setups to provide them with frost oil would also be neat

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+/* pink rune, makes people slightly happier after walking on it*/
+
+
+/obj/item/slimecross/warping/pink
```
pretty nice rune, I can see this being used

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+	. = ..()
+	if(istype(crossing,/mob/living/carbon/human))
+		playsound(rune_turf, "sound/weapons/thudswoosh.ogg", 50, TRUE)
+		SEND_SIGNAL(crossing, COMSIG_ADD_MOOD_EVENT,"jolly", /datum/mood_event/jolly)
```
can't you just use the hug moodlet (or even the betterhug or besthug moodlets)?

Caldony replied:
I thought about using the hug moodlet but it isn't that high, and the betterhug and besthug moodlet run into a runtime when not used by anyone which it technically isn't since it gives the mood directly without the action

ATH1909 replied:
can't you just modify the hug and betterhug moodlets to NOT runtime when used/applied by an object or the like?

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+
+/*Gold rune : Turn things over it to gold, completely fucks over economy */
+
+/obj/item/slimecross/warping/gold
```
aw, you didn't go with my summoning rune idea from my last PR?

well, this rune still has the issues that it had before, including the lore issue of the gold material falling under the domain of metal slime cores, NOT gold slime cores

this rune might also run into the various problems that metalgen did

Caldony replied:
In all honesty I'm probably going to change it as it's way too likely to fuck up the economy balance eventually, and can easily be abused with massive item stacks made from less expensive materials in general

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+/*Adamantine rune, will spawn ores depending on the mineral rocks surrounding it. Here to make miners do their job even less.  */
+
+
+/obj/item/slimecross/warping/adamantine
```
the grievances that I had from last PR still apply here

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+
+
+/obj/effect/warped_rune/adamantinespace/proc/auto_mining()
+	for(var/turf/closed/mineral/ores in range(7,rune_turf)) //the range is pretty big to at least try to rival miners and their plasma cutters.
```
the large range is actually a detriment, as in order to cherry pick the ores this thing will harvest, you're gonna have to remove all other kinds of ore from this range

Caldony replied:
Because it will do that for each mineral in range, it will still copy ALL ores in range, not just the first one it comes across, so if you have 10 diamond vein in range it will produce 10 diamond ores for exemple.

ATH1909 replied:
ohhhhhhhhhh

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+/obj/item/slimecross/warping/lightpink
+	colour = "light pink"
+	runepath = /obj/effect/warped_rune/lightpinkspace
+	effect_desc = "draws a rune that will repair the soul of a humanoid corpse in the hope of bringing them back to life."
```
specify that the corpse has to be soulless, and maybe specify that you're replacing the soul instead of repairing it

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+		return
+
+	for(var/mob/living/carbon/human/host in rune_turf)
+		if(!host.getorgan(/obj/item/organ/brain) || host.key || host.get_ghost(FALSE, TRUE)) //checks if the ghost and brain's there
```
(with regards to the call to get_ghost): I mean, if they can't reenter their corpse, I don't see why we can't let someone else steal their dead body

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+/obj/effect/warped_rune/lightpinkspace
+	icon_state = "necro_rune"
+	desc = "Souls are like any other material, You just have to find the right place to manufacture them."
+	max_cooldown = 100
```
might want to increase this cooldown to, like, a minute, to prevent ghost ping spam (and to keep xenobio from being able to pump out living humanoids TOO quickly)

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+			return
+
+		else
+			to_chat(user, "<span class='warning'>The rune failed! Maybe you should try again later.</span>")
```
dust the corpse if you can't find any ghost candidates for it, so that leaving a corpse on a rune (or leaving MULTIPLE corpses on MULTIPLE runes) won't ping ghosts every minute until they select the "never" option for that prompt

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+			host.key = ghost.key
+			host.suiciding = 0 //turns off the suicide var just in case
+			host.revive(full_heal = TRUE, admin_revive = TRUE) //might as well go all the way
+			to_chat(host, "<span class='warning'>You may wear the skin of someone else, but you know who and what you are. Pretend to be the original owner of this body as best as you can.</span>")
```
might want to specify if they retain the memories from their previous life

maybe make impersonating the person they've been revived as a proper objective?

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+			host.suiciding = 0 //turns off the suicide var just in case
+			host.revive(full_heal = TRUE, admin_revive = TRUE) //might as well go all the way
+			to_chat(host, "<span class='warning'>You may wear the skin of someone else, but you know who and what you are. Pretend to be the original owner of this body as best as you can.</span>")
+			to_chat(user, "<span class='notice'>[host.name] is slowly getting back up with an empty look in [host.p_their()] eyes. It...worked?</span>")
```
```suggestion
			to_chat(user, "<span class='notice'>[host.name] is slowly getting back up with a strange look in [host.p_their()] eyes. It... worked?</span>")
```
might want to also add a special message for if the corpse is of a race that doesn't HAVE eyes

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+		to_chat(crossing, "<span class='warning'>The rune needs a little more time before processing your DNA!</span>")
+		return
+
+	if(!istype(crossing,/mob/living/carbon/human))
```
also check to see if they have a species/dna, just in case

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+/* black space rune : will swap out the species of the two next person walking on the rune  */
+
+
+/obj/item/slimecross/warping/black
```
really cool rune

note that I think that this rune can turn people into plasmamen, but I don't think that that's TOO much of a problem (it's much harder to use than, say, plasmaman mutation toxin)

ATH1909 replied:
if you DO want to blacklist plasmamen and the like from this, I'm pretty sure that there's a list of races that are blacklisted from changeling transformation stings and the like somewhere (maybe it's a variable on the races themselves? idk, you'll have to codedive to check)

XDTM replied:
it's a species trait, NOMUTSTING i think

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+/*Acts as a bunker making anything into the rune immune to outside explosions and gas leaks. doesn't work if the epicenter of the explosion is IN the rune */
+
+
+/obj/item/slimecross/warping/oil
```
could you maybe make this protect a wider area, so you can use this to "ward" departments against explosions/plasma fires?

ATH1909 replied:
cool rune, though

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+	invisibility = TRUE //no one can see those
+	anchored = TRUE
+	explosion_block = INFINITY
+	CanAtmosPass = ATMOS_PASS_NO //will also try to stop gas from exiting the rune to keep the pressure and air of the tile.Doesn't always work
```
might want to ensure that you can't accidentally suffocate departments with these tiles by blocking the atmos vents/entrances with them

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+/* Used to teleport anything over it to a unique room similar to hilbert's hotel.*/
+
+
+/obj/item/slimecross/warping/rainbow
```
cool rune

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+///anyone stepping on the "exit rune" will be teleported back on the original rune
+/obj/effect/warped_room_exit/attack_hand(mob/living/user)
+	. = ..()
+	for(var/mob/living/carbon/human/customer in get_turf(src)) //only people in the customers list are allowed to leave, no bluespace bodybag smuggling in MY hotel
```
I'm very confused by this comment

XDTM replied:
probably comes from hilbert hotel code

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+		customer_check(customer, TRUE)
+
+
+///teleport all customers in the room back to the rune when the rune is destroyed if no customers have left
```
are you sure that this works?

ATH1909 commented on tgstation.dme:
```diff
@@ -2821,6 +2821,7 @@
 #include "code\modules\research\xenobiology\crossbreeding\reproductive.dm"
 #include "code\modules\research\xenobiology\crossbreeding\selfsustaining.dm"
 #include "code\modules\research\xenobiology\crossbreeding\stabilized.dm"
+#include "code\modules\research\xenobiology\crossbreeding\warping.dm"
```
well, that's the end of my quick review

let's see how many comments I ended up ma- oh god

Caldony commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+
+
+///runes can also be deleted by bluespace crystals relatively fast as an alternative to cleaning them.
+/obj/effect/warped_rune/attackby(obj/item/used_item, mob/user)
```
yeah it's in the after_attack of the warped crossbreed themselves

Rohesie replied:
Have you tried this on different intents? Doesn't it do an attack animation on harm?

XDTM commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+
+	for(var/mob/living/on_fire in rune_turf)
+		on_fire.adjust_fire_stacks(20)
+		on_fire.IgniteMob()
```
it seems to have been removed after my review

XDTM commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+/obj/effect/warped_rune/yellowspace/proc/electrivore(obj/recharged)
+	if(recharged.get_cell()) //check if the item has a cell
+		var/obj/item/stock_parts/cell/battery = recharged.get_cell()
+		if(battery.charge <! battery.maxcharge) //don't charge if the battery is full
```
99% a typo, it only compiles because it interprets it as battery.charge < !battery.maxcharge

XDTM commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+/*  Yellow rune space acts as an infinite generator, works without power and anywhere, recharges the APC of the room it's in and any battery fueled things. */
+
+
+/obj/item/slimecross/warping/yellow
```
not in this current state though, it loops through all atoms in the area

ATH1909 replied:
oh, I thought it just looped through all of the atoms ON THE RUNE

Caldony replied:
A little bit of a late response, but no it does recharge both the APC and items ON the rune, the proc is called twice, once for the APC in the rune area and once for any items on the rune. I may lower the cooldown. If another item is in the area but not on the rune it won't be checked at all.On another note the gold rune has been reworked so there's that.

XDTM commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+/* makes a rune that absorb food, whenever someone step on the rune the nutrition come back to them, not all of it of course.*/
+
+
+/obj/item/slimecross/warping/silver
```
where's the interaction, just feed it with more silver slimes

Caldony commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+/*Metal rune : makes an invisible wall. actually I lied, the rune is the wall.*/
+
+
+/obj/item/slimecross/warping/metal
```
nope, it blocks out any projectile from passing by it, although it will act as if the rune is hit which I guess it technically is, the rune won't take any damage from it. It's basically a see through wall.

Caldony commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1135 @@
+///it makes people that step on the tile very cold.
+/obj/effect/warped_rune/darkbluespace/proc/cold_tile()
+	for(var/mob/living/carbon/human in rune_turf)
+		human.adjust_bodytemperature(-1000) //Not enough to crit anyone not already weak to cold, might need serious rebalance if cold damage is reworked.
```
I never saw the temperature actually go under -1000° even if you continuously stay on the rune because the temperature of the body just adapts incredibly fast for really low temperatures, you'd need to already be really fucking cold to get to absolute zero

Caldony committed:
Fixes shitcode I think

Updates a few comment to be more relevent to the rune
Silver now works better I guess
cerulean now works slightly differently and has a proper cooldown along with timers shenanigans.
Sepia will now properly reapply the slow debuff.
Fixed a few typos codewise
I'm tired so I hope I didn't slip more shitcode in

Caldony committed:
Fixes the sepia timer

Fixes the sepia timer by adding a UNIQUE and OVERRIDE trait to the timer so it can reset the timer even with multiple people walking on it.

Caldony committed:
Makes the cerulean timer less shit

use UNIQUE and OVERRIDE for the cerulean timer too so we get rid of useless vars.

ATH1909 commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1147 @@
+	var/obj/item/stack/sheet/xeno_resin/resin
+
+
+/obj/item/stack/sheet/xeno_resin
```
it'd be nice if you could make it so that (real) xenomorphs can hold these resin sheets, sort of like how they can hold facehuggers and the like

Caldony committed:
Reworks the gold rune completely

Communism is no more!
The gold rune will now absorb hostile simple mobs into itself (with the exception of megafauna) and whoever next step on the rune will spawn all the mobs absorbed by the rune and give them whoever stepped on the rune as a priority target. 

They will spawn ANGERY though so once they lose their target I wouldn't count on them not eating your ass. One rune can contain up to 10 mobs, making for a VERY lethal trap considering how fast mobs tend to react if they are given a target in advance depending on which mobs the rune absorbed.

Caldony committed:
Travis you whore give me back my green check

changes a few minor balance things
also give a constant teleportation effect to whoever get put on fire by the orange rune. The teleportation isn't instant however and will act every 7 second (5 was too fast and 10 too long so I decided to go for 7)

Caldony committed:
Properly calls parent on crossed()

Caldony committed:
rebalances the gold rune a little

make the description of the effect say "creature" instead of "animals" cause it makes more sense
the rune can now only store 5 mobs instead of 10 because holy fuck 10 is a lot of mob to spawn at the same time

Caldony committed:
Lightpink fix (#6)

* adds the ignore poll for the lightpink rune

* fixes the lightpink rune and adds a few things

The revived person will now get "soul directives" in their memory to follow, 
also fixes a bug where people wouldn't get resurrected properly and adds a bigger cooldown I think

Caldony committed:
Rebalances the orange rune

the orange rune will no longer teleport dead bodies or if the rune has been deleted

Caldony committed:
Replace most lists with lazy lists

what the title says, some temporary lists that only last for a proc aren't set to lazy list for obvious reason.
The gold rune, the oil rune, the cerulean rune and the rainbow rune now all have their list set to null when empty one way or another.

Caldony committed:
remove useless qdel() in the purple rune

The purple rune had the stack being qdeleted when the stack reached 0 even though use() already does that, so I'm removing that
Also fixes the problem where one plastic sheet on the rune would make it impossible to create suture out of cloth

Caldony committed:
Runes now do things when drawn

Basically if you were on the rune at the moment you drew the rune it wouldn't affect you as it technically only affect people CROSSING over the rune instead of who's over it.that fixes that for most runes like redspace, ceruleanspace and sepiaspace

Caldony committed:
"Soul directives" will now delete the previous owner memory

Lightpink will now empty the previous owner of the body's memory and replace it with the soul directive, this include any previous objectives the original owner of the body may have had.
Also yellow rune now recharge every 5 seconds instead of 10

Caldony committed:
Boost up the speed of the gold rune mobs

The mob spawned by the gold rune will now have a move_delay of 1, making them slightly slower than bees and one of the fastest hostile mobs to exist. Making them much more fearsome.

Caldony committed:
Yellow runes now recharge borgs and mechs!

I had to add a new proc for the borgs since they aren't compatible with the chad obj.get_cell()

ExcessiveUseOfCobblestone commented:
why is ever commit webbited?

Caldony commented:
> 
> 
> why is every commit webbited?

For some reason the thing I use to push my commit has been pretty fucky recently so I had to update most of warping.dm by uploading the file directly unto github everytime. I did test locally each commit. It's just that pushing said commit has been a fucking nightmare with my computer.

caluany committed:
minor grammatical fixes

Of course commit machine isn't broke anymore when I say it is

Anyway, just change a few grammatical things. Orange rune will also now keep teleporting you even if the rune is destroyed. Plasma rune produce 5 unit every 5 seconds instead of 50 every 30 seconds.

caluany committed:
Revert "minor grammatical fixes"

This reverts commit 05aed7bd5853e0b1423531ef2fcb9bfdb2b09b38.

caluany committed:
Another test commit

This one better work I swear to god

Rohesie commented on code/__DEFINES/layers.dm:
```diff
@@ -89,11 +90,12 @@
 #define EMISSIVE_BLOCKER_LAYER 12
 #define EMISSIVE_BLOCKER_RENDER_TARGET "*EMISSIVE_BLOCKER_PLANE"
 
+#define CHAT_LAYER 12.1
```
Is this an intended change?

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1227 @@
+
+/obj/item/slimecross/warping/Initialize()
+	. = ..()
+	desc = "It just won't stay in place. it has [warp_charge] charge left"
```
Maybe add the charge part to the examine proc?

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1227 @@
+	if(!proximity)
+		return
+
+	if(isturf(target) && locate(/obj/effect/warped_rune) in target) //check if the target is a floor and if there's a rune on said floor
+		to_chat(user, "<span class='warning'>There is already a bluespace rune here!</span>")
+		return
+
+	if(istype(target, runepath))       //checks if the target is a rune and then if you can store it
+		if(warp_charge >= max_charge)
+			to_chat(user, "<span class='warning'>[src] is already full!</span>")
+			return
+
+		if(do_after(user, storing_time,target = target) && warp_charge < max_charge)
+			to_chat(user, "<span class='notice'>You store the rune in [src].</span>")
+			qdel(target)
+			warp_charge++
+			desc = "It just won't stay in place. it has [warp_charge] charge left"
+			return
+
+	if(!istype(target,/turf/open))
+		return
+
+	if(isspaceturf(target))
```
Do redefined procs instead of long istype checks.
Things such as `warping_slime_act()` works.

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1227 @@
+
+	if(!length(enter_rune.customer_list) && !locate(enter_rune) in exit_turf) //deletes the room if the rune doesn't exist anymore and all customers have left
+		qdel(enter_rune.room_reservation)
+		qdel(src)
+		return ..()
```
deletion will end the proc call
Are you sure you want to call the parent on a deleted object?

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1227 @@
+
+
+///Will delete the room when the rune is destroyed if no customer is left in the room.
+/obj/effect/warped_rune/rainbowspace/Destroy()
+	if(!length(customer_list))
+		qdel(room_reservation)
+	return ..()
```
Clean the references to `customer_list`, `room_reservation` and `rune_room` on deletion.
If they can be destroyed before the rune, then register to their destruction and clean them up as well.

Rohesie replied:
Also, since this is a lazy list, use `LAZYLEN` instead of `length`
It's quite the same, but it indicates it being one.

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1227 @@
+
+
+///here to check if anyone's being transported in or out of the room with the user.
+/obj/effect/warped_rune/rainbowspace/proc/customer_check(atom/person_checked ,smuggle_in)
```
```suggestion
/obj/effect/warped_rune/rainbowspace/proc/customer_check(atom/person_checked, smuggle_in)
```

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1227 @@
+	move_resist = INFINITY
+	anchored = TRUE
+	///where the rune will teleport you back.
+	var/turf/exit_turf
+	///rune linked to the exit rune
+	var/obj/effect/warped_rune/rainbowspace/enter_rune
```
Clean these variables on `Destroy()` and register on their destruction in case they may be destroyed before the effect is.

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1227 @@
+		LAZYADD(bunker_list, bunker_wall)
+
+
+/obj/effect/warped_rune/oilspace/Destroy()
+	for(var/obj/effect/oilspace_bunker/bunker_wall in range(1,rune_turf))
+		if(bunker_wall in bunker_list)
+			LAZYREMOVE(bunker_list, bunker_wall)
+			qdel(bunker_wall)
+			return ..()
```
`return ..()` is not properly indented.
Clean `bunker_list` on destruction.

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1227 @@
+
+
+///will swap the species of the first two human or human subset that walk on the rune
+/obj/effect/warped_rune/blackspace/Crossed(atom/movable/crossing)
+	. = ..()
+	if(cooldown > world.time) //here to avoid spam/lag
+		to_chat(crossing, "<span class='warning'>The rune needs a little more time before processing your DNA!</span>")
+		return
```
Won't this spam ghosts and the likes? Put this after the istype check.

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1227 @@
+		return
+
+	var/mob/living/carbon/human/crosser = crossing
+	if(!stepped_on)
+		first_person = crosser
+		stepped_on = TRUE
+		return
```
Register on their destruction, not to keep a destroyed mob reference.

Rohesie replied:
Same for the `second_person`

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1227 @@
+/obj/effect/warped_rune/blackspace/Initialize()
+	. = ..()
+	cooldown = 0 //doesn't start on cooldown like most runes
+	for(var/atom/movable/crossing in rune_turf)
+		Crossed(crossing)
```
Call a custom proc also called by `Crossed()`
Don't call `Crossed()` directly.

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1227 @@
+	colour = "light pink"
+	runepath = /obj/effect/warped_rune/lightpinkspace
+	effect_desc = "draws a rune that will attempt to repair a soulless humanoid corpse in the hope of bringing them back to life."
+	drawing_time = 100
```
```suggestion
	drawing_time = 10 SECONDS
```

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1227 @@
+/obj/effect/warped_rune/lightpinkspace
+	icon_state = "necro_rune"
+	desc = "Souls are like any other material, you just have to find the right place to manufacture them."
+	max_cooldown = 300
```
```suggestion
	max_cooldown = 30 SECONDS
```

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1227 @@
+		to_chat(user, "<span class='warning'>The rune is still charging!</span>")
+		return
+
+	for(var/mob/living/carbon/human/host in rune_turf)
```
Are you sure you want a `for` loop here, instead of a `locate` proc?

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1227 @@
+	icon_state = "mining_rune"
+
+
+/obj/effect/warped_rune/adamantinespace/Initialize()
+	. = ..()
+	START_PROCESSING(SSprocessing, src)
+
+
+/obj/effect/warped_rune/adamantinespace/process()
+	if(cooldown < world.time)
+		cooldown = world.time + max_cooldown
+		auto_mining()
```
Does this need to be a continuous process instead of activated on demand?
Unless this is moving, it won't need to scan its surroundings after the first time, and if it's moving the user can keep activating it as needed, no?

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1227 @@
+	merge_type = /obj/item/stack/sheet/xeno_resin
+
+
+/obj/effect/warped_rune/greenspace/Initialize()
+	. = ..()
+	START_PROCESSING(SSprocessing, src)
+
+
+/obj/effect/warped_rune/greenspace/process()
+	if(cooldown < world.time)
+		cooldown = world.time + max_cooldown
+		transmute_resin()
```
Could you do this on rune activation instead of a continuous process?

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1227 @@
+
+/obj/effect/warped_rune/greenspace/proc/transmute_resin()
+	for(var/obj/item/stack/sheet/mineral/plasma/plasma_sheet in rune_turf)
+		resin = new(rune_turf)
```
Why are you storing the instantiated object?

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1227 @@
+		Crossed(enraged)
+
+
+///boost up the unarmed damage of the person currently on the tile.
+/obj/effect/warped_rune/redspace/Crossed(atom/movable/crossing)
+	. = ..()
+	if(!istype(crossing,/mob/living/carbon/human))
+		return
+
+	var/mob/living/carbon/human/enraged = crossing
+	enraged.add_overlay(hate_fire)
+	RegisterSignal(enraged, COMSIG_HUMAN_EARLY_UNARMED_ATTACK, .proc/genetic_punch)
+	to_chat(enraged, "<span class='warning'>You feel the hate of a thousand souls culminating into your fists!</span>")
+
```
Register to `COMSIG_MOVABLE_UNCROSSED` to remove this, instead of checking again in `Uncrossed()`

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1227 @@
+	desc = "Progress is made through adversity, power is obtained through violence"
+	icon_state = "rage_rune"
+	///fire overlay appearing behind someone when they are on the rune
+	var/mutable_appearance/hate_fire
```
Could you turn this into a static visual object?

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1227 @@
+	colour = "red"
+	runepath = /obj/effect/warped_rune/redspace
+	effect_desc = "Draws a rune giving your fists the ability to hurt the very soul of whoever you punch, healing you in the process."
+	drawing_time = 30
```
```suggestion
	drawing_time = 3 SECONDS
```

Rohesie commented on code/modules/research/xenobiology/crossbreeding/warping.dm:
```diff
@@ -0,0 +1,1227 @@
+
+
+///colors whatever steps on the rune randomly
+/obj/effect/warped_rune/pyritespace/Crossed(atom/movable/colored)
+	. = ..()
+	colored.color = rgb(rand(0,255),rand(0,255),rand(0,255))
```
This will color everything, including ghosts and whatnot.
Add some limits here.

caluany committed:
Fixes a few things

change the examine and the way runes are drawn.

use SECONDS for most time related things.

green is now activable by using it with empty hands.

caluany committed:
Accidentally called the parent of an orphan

caluany committed:
slightly change the yellow rune

the check for recharging the mech and recharging items were different even though they are both under the obj system so they both use the same proc.

ExcessiveUseOfCobblestone commented:
can we please chill with crossbreeding if someone merges this?

the only department that can get pseudo-wizard powers, raise an unsubvertable army (2 ways), get near-infinite charge batteries, and I believe

ensure no one dies from crit around the user
ignore slowdown of items
tracer rewind ability?
turn hostile mobs friendly
turn mobs into PCs

and probably a lot more. The only limiting factor is RNG, part upgrades (which these people have direct access to build and rush), and time to process.

I mean I know chemistry does this kinda junk too where it can answer 500 problems but there's not much more to the department, but that is bonkers.

caluany committed:
slightly change formatting

just give the spawning and absorbing of the crossbreed their own proc. Mainly here to look cool

also fixed the silver examine proc

also aslo, fixed a problem with the rune being able to spawn more than one rune at a time on the same turf if they were different rune.

Rohesie commented:
Agreed with Cobby. We might need to review the department's capabilities and remove the worst offending ones.

Let me know when this is ready for another round of code review.

Caldony commented:
I agree actually, Xenobio may need a serious overhaul in general and hiding most of these near game breaking artifacts behind a timewall instead of a clear risk versus reward scenario doesn't feel like enough. 

While I do love a lot of the wacky effects and the near magical powers of the crossbreed and their versatility, xenobio doesn't feel like a rightful owner of all that power, no jobs would deserve that much power

I'll try to fix what I can but I won't hold it against anyone if they decide to close it with how hard to predict and balance some runes are.

Caldony commented:
I'm going to close this PR, I honestly just don't have the heart to work on it, it's a lot of work for something I don't care for much anymore.

I'm sorry for everyone who wanted to see the crossbreeds updated or that some of those ideas were cool. I might release PR for some of easier to balance crossbreeds later or I might not. You'r free to take a bunch of my code to port somewhere else if you want as long as you credit me, or don't, not like I can sue you.

I'd still like to thank @Rohesie  and @ATH1909 for reviewing this monster of a PR and helping me improve my code bit by bit, your efforts weren't completely wasted, probably.
